api/common/interfaces.go

```diff
@@ -647,91 +647,152 @@ func getSubscribers(s Serverer, params map[string]interface{}) map[string]interf
+	if err != nil {
+		return respPacking(INTERNAL_ERROR, err.Error())
 	}
 
-	subscribers, err := chain.DefaultLedger.Store.GetSubscribers(topic, uint32(bucket))
+	meta, expiresAt, err := chain.DefaultLedger.Store.GetSubscription(topic, uint32(bucket), pubKey, identifier)
 	if err != nil {
 		return respPacking(INTERNAL_ERROR, err.Error())
 	}
-	return respPacking(SUCCESS, subscribers)
+	return respPacking(SUCCESS, struct {
+		Meta      string `json:"meta"`
+		ExpiresAt uint32 `json:"expiresAt"`
+	}{
+		meta,
+		expiresAt,
+	})
 }
 
-// getAsset get subscribers by topic
-// params: {"assetid":<id>}
+// getSubscribers get subscribers by topic
+// params: {"topic":<topic>, "bucket":<bucket>, "offset":<offset>, "limit":<limit>}
 // return: {"resultOrData":<result>|<error data>, "error":<errcode>}
-func getAsset(s Serverer, params map[string]interface{}) map[string]interface{} {
-	if len(params) < 1 {
-		return respPacking(INVALID_PARAMS, "length of params is less than 1")
+func getSubscribers(s Serverer, params map[string]interface{}) map[string]interface{} {
+	if len(params) < 2 {
+		return respPacking(INVALID_PARAMS, "length of params is less than 2")
 	}
 
-	str, ok := params["assetid"].(string)
+	topic, ok := params["topic"].(string)
 	if !ok {
-		return respPacking(INVALID_PARAMS, "asset ID should be a string")
+		return respPacking(INVALID_PARAMS, "topic should be a string")
 	}
 
-	hexAssetID, err := common.HexStringToBytes(str)
-	if err != nil {
-		return respPacking(INVALID_PARAMS, err.Error())
+	var bucket float64
+	if _, ok := params["bucket"]; ok {
+		bucket, ok = params["bucket"].(float64)
+		if !ok {
+			return respPacking(INVALID_PARAMS, "bucket should be a float64")
+		}
 	}
 
-	assetID, err := common.Uint256ParseFromBytes(hexAssetID)
-	if err != nil {
-		return respPacking(INVALID_PARAMS, err.Error())
+	var offset float64
+	if _, ok := params["offset"]; ok {
+		bucket, ok = params["offset"].(float64)
+		if !ok {
+			return respPacking(INVALID_PARAMS, "offset should be a float64")
+		}
 	}
 
-	name, symbol, totalSupply, precision, err := chain.DefaultLedger.Store.GetAsset(assetID)
+	var limit float64
+	if _, ok := params["limit"]; ok {
+		bucket, ok = params["limit"].(float64)
+		if !ok {
+			return respPacking(INVALID_PARAMS, "limit should be a float64")
+		}
+	} else {
+		limit = 1000
+	}
+
+	localNode, err := s.GetNetNode()
 	if err != nil {
 		return respPacking(INTERNAL_ERROR, err.Error())
 	}
 
-	ret := map[string]interface{}{
-		"name":        name,
-		"symbol":      symbol,
-		"totalSupply": totalSupply.String(),
-		"precision":   precision,
+	subscribers, err := chain.DefaultLedger.Store.GetSubscribers(topic, uint32(bucket), uint32(offset), uint32(limit))
+	if err != nil {
+		return respPacking(INTERNAL_ERROR, err.Error())
 	}
-
-	return respPacking(SUCCESS, ret)
+	return respPacking(SUCCESS, map[string][]string{
+		"subscribers":         subscribers,
+		"subscribersInTxPool": localNode.GetTxnPool().GetSubscribers(),
```

Comment:
We should return subscribers in txpool only for the specified topic. In the current network, all subscribers in pool constantly contains ~30k results (30M in size)...

Reply:
oh, you're right, I missed that somehow

Reply:
also prob should be made optional